### PR TITLE
Multi-turn conversation dataset

### PR DESCRIPTION
**Pull request by jbrry**
Based on the previous conversation in https://github.com/ibm-granite/dolomite-engine/pull/143 (@mayank31398 @murthyrudra), I created another branch/PR that just adds this single dataset.

### Background

The dataset supports loading local .jsonl files which have the format:
```json
{"messages":[{"content":"Aster is a chatbot who answers questions with rhymes.","role":"system"},{"content":"Where did chocolate originate?","role":"user"},{"content":"Chocolate is 4000 years old\/Mexico is where it was first sold","role":"assistant"},{"content":"Where was milk chocolate invented?","role":"user"},{"content":"Switzerland was the first to add milk\/To make their chocolate smooth as silk","role":"assistant"},{"content":"What are some good desserts that use chocolate?","role":"user"},{"content":"Pie, tart, cookies, and cake\/Chocolate is great to bake","role":"assistant"}]}
```

Due to the interleaving messages between `user` and `assistant` a custom function [construct_input_output_from_chat_format](https://github.com/jbrry/dolomite-engine/blob/46f7be76486299fa3ad877d9acc52d60a92a092a/dolomite_engine/data/huggingface_chat_format_dataset.py#L78) is used to create the input/output lists. It builds the input and output using chat templates, and for the output it sets the `labels_mask_value` everywhere except the assistant responses. As far as I can see, no changes are needed to the collate function, as this is equivalent to having all tokens in the labels as -100 except the negative/positive strings in the `sst2` example, but the attention mask will still include the conversation tokens, and just ignore length-related padding.

The config file I am using is here:
```yaml
datasets:
  - class_name: HuggingFaceChatFormatDataset
    class_args:
      data_path: /path/to/input_data
      eval_steps: 10
      chat_template_tag: "default"
    data_name: chat
    output_format: " __output__"
    data_sampling_ratio: 1
    max_input_tokens: 1024
    max_output_tokens: 1024

model_args:
  model_name: bigscience/bloom-560m
  model_class: AutoModelForCausalLM

tuning_args:
  tuning_method: full_finetuning

save_args:
  save_path: /dccstor/jbarry_storage/work_projects/bleeding_edge/model_output
  save_interval: 1000

training_parameters:
  num_training_steps: 1500
  eval_interval: 100
  micro_batch_size: 2
  gradient_accumulation_steps: 2

optimizer_args:
  class_name: TorchAdamW
  class_args:
    lr: 1e-4
    weight_decay: 0.1
    betas:
      - 0.9
      - 0.95
    eps: 1e-10

mixed_precision_args:
  dtype: fp32

lr_scheduler_args:
  lr_decay_style: cosine

distributed_args:
  distributed_backend: torch
  gradient_checkpointing_method: block
```

Some of the values are currently redundant, e.g. `max_input_tokens`

### CODE REVIEW

**Comment by jbrry** [comment]
There is also a difference to other datasets currently: we do not truncate to `max_length` afterwards and append EOS token like [here](https://github.com/jbrry/dolomite-engine/blob/46f7be76486299fa3ad877d9acc52d60a92a092a/dolomite_engine/data/base.py#L112).

The main reason is that it wouldn't be necessary to put the last element as EOS if it is a `user` message.

Currently, what this looks like for `assistant` messages is:
`'<rest of assistant message> ... On the weekends, the market hosts various'`

But if we ensured that the `assistant` always ends with EOS it would look like:
`'<rest of assistant message> ... On the weekends, the market hosts </s>'`

I can implement the second option if necessary.
